draft/XXXX-missing-values-in-typed-cols/README.md

```diff
@@ -0,0 +1,44 @@
+---
+RFC: XXXX
+Author: Jack Wilburn
+Created: 2020-10-26
+Last-Modified: 2020-10-26
+---
+
+# RFC XXXX: Missing Values In Typed Columns
+
+As we move to having typed data in Multinet, it is important to handle missing data elegantly, even with typing. Currently, if a column has any amount of missing data, we assume that the data is a "label" type. Since the "label" type is essentially a catch-all for data that we can't provide a meaningful type for, it's important that if there is a semantically meaningful type, we provide it to visualization applications. This must, also, be true in the case where we have mostly complete data with just a few missing rows. Without this information, the client applications cannot adequately choose a visualization for these columns.
+
+There will likely be some client- and server-side considerations for handling missing data. This will include correctly parsing data and when reading it into the ArangoDB types (or we may choose to leave them as strings in the db). Types that we define, will be referenced as semantic types, and native ArangoDB types will be referenced as Arango types
+
+## Motivation
+
+We now have a motivation to supply data schema information to visualization applications, UpSet. Upset requires set membership information and other data types to generate its visualization. We can also utilize this type information in the other visualization applications that support multinet (MultiLink and MultiMatrix).
+
+Since we now need to pass semantic typing into these applications, knowing what the types are, even if there is missing data, is essential. As such, we'll need to modify our typing logic to handle missing data. This stretches from the frontend, where we provide type suggestions, to the backend, where we store the type information and disseminate it to the vis applications.
+
+## Proposals
+
+We should provide type information for columns that are of one type, but that have missing values. There are at least 3 places that we should support this paradigm:
+
+1. `multinet-client` on data preview/type suggestion.
+2. `multinet-server` on data upload and data fetching.
+3. `multinetjs` on data upload and data fetching.
+
+For the type suggester, there is a code chunk in the [Reference Implementation](#reference-implementation) section that suggests a way that we can check for the type of a column, even when data is missing.
+
+For the other repos, `multinet-server` and `multinetjs`, the considerations are slightly different. We'll need to think about how to store and serve data with missing values, and how to disseminate the type information given that there are missing values. Additionally, an open question is, should we tell client applications to expect some missing data in typed columns? This could help when rendering visualizations.
```

Comment:
My thoughts on this:
1. Storing/serving missing values is easy: this is the exact semantics of the JSON/JavaScript `null` value, so we simply store `null` in the DB, and serve it upon retrieval just like any other value.
2. Whether to indicate missing values in a column is a slightly harder question, but I'm leaning towards "yes". If we're considering the type of a column to be a special kind of column metadata entry, then we could also have a boolean flag called `complete` or something like that (we will need to think hard about what term is best) as well. As with all other metadata, vis apps would be free to make use of or ignore this flag at its discretion.

Reply:
I'm not necessarily against storing the "completeness" of a column, but is there an example use case for this?

Reply:
I don't think we have a concrete use case now, but the idea would be to signal to apps that they may need to check for `null` when doing computations, etc. If setting a metadata flag to indicate this has problems associated with it, we don't have to do it, but I can see it streamlining vis app code. For instance, check to see if `!metadata.complete`, and run an imputation or filtering step on the data before proceeding.

Not a big deal if we don't include it, or if we do so after a concrete need arises.

Reply:
I think the client should be deciding if there are `null` values in the dataset instead of relying on the server for that info for now. Unless that become a bottleneck, at which point we can revisit this

Reply:
You're right. Even if the type indicates there are no missing values, I don't expect that to make coding vis apps any easier: you would still need to check for `null` in some branch of the code where the type indicates there are missing values.

Together with @AlmightyYakob's argument that we don't have a valid need for this right now, that means we shouldn't support this now.

Go ahead and update the paragraph to reflect that, plus my comments about using `null` values above.

Reply:
0d938f8